pychromecast/controllers/multizone.py

```diff
@@ -2,27 +2,173 @@
 Controller to monitor audio group members.
 """
 import logging
+from uuid import UUID  # noqa: F401,E501 pylint: disable=unused-import
+import threading
 
 from . import BaseController
+from .media import MediaController, MediaStatus  # noqa: F401,E501 pylint: disable=unused-import
+from ..socket_client import (
+    CONNECTION_STATUS_CONNECTED, CONNECTION_STATUS_DISCONNECTED,
+    CONNECTION_STATUS_LOST)
 
 _LOGGER = logging.getLogger(__name__)
 
 MESSAGE_TYPE = "type"
 MULTIZONE_NAMESPACE = "urn:x-cast:com.google.cast.multizone"
+TYPE_CASTING_GROUPS = "CASTING_GROUPS"
 TYPE_DEVICE_ADDED = "DEVICE_ADDED"
 TYPE_DEVICE_UPDATED = "DEVICE_UPDATED"
 TYPE_DEVICE_REMOVED = "DEVICE_REMOVED"
+TYPE_GET_CASTING_GROUPS = "GET_CASTING_GROUPS"
 TYPE_GET_STATUS = "GET_STATUS"
 TYPE_MULTIZONE_STATUS = "MULTIZONE_STATUS"
+TYPE_SESSION_UPDATED = "PLAYBACK_SESSION_UPDATED"
+
+
+class MultizoneManager:
+    """ Manage audio groups. """
+    def __init__(self):
+        # Protect self._casts because it will be accessed from callbacks from
+        # the casts' socket_client thread
+        self._casts = {}
+        self._groups = {}
+        self._lock = threading.Lock()
+
+    def add_multizone(self, group_cast):
+        """ Start managing a group """
+        lock = self._lock
```

Comment:
oh no, locks ! Can we do without?

Reply:
We need to protect self._casts because it will be accessed from callbacks from the casts' socket_client threads.
It could be modified such that the lock is only held when accessing the dict though, or remove the lock and rely on the GIL:
> This simplifies the CPython implementation by making the object model (including critical built-in types such as dict) implicitly safe against concurrent access.